### PR TITLE
feat(AwsVllmComponent): Add AWS Cognito as auth service

### PR DESCRIPTION
**Pull request by bramelfrink**
Addresses https://github.com/DataChefHQ/damavand/issues/26

The assumption that I'm making in this PR is that: A [Cognito user pool](https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools.html) is created elsewhere; so it is not the responsibility of the `AwsVllmComponent` to do so.

To do:
- [ ] test this with a real application / user pool.

### CODE REVIEW

**Comment by kkiani** [comment]
Thanks, @bramelfrink, for picking this up. One thing I forgot to mention while writing the issue—and was about to update just before receiving your PR—is that the APIGateway is supposed to provide an interface fully compatible with the [OpenAI Chat Completion API Model](https://platform.openai.com/docs/api-reference/chat/create). This means the end user should be able to make a request using a standard API key. I am not sure if AWS Cognito can provide such functionality. However, a user authentication feature might be a nice addition if we eventually decide to build a Chat UI on top of this, which is not currently on the roadmap.

**Comment by kkiani** [comment]
The APIGateway does have a functionality for creating API Keys but one thing that is important is to be able to create API Keys after api deployment without a need for `pulumi up`.

**Comment by bramelfrink** [comment]
@kkiani The creation dynamic creation of API keys should not be part of this application IMO. What you could do here is define (configurable) usage plans, as a key needs to be linked to a usage plan.

The main question is: do we create the usage plans here, or should that go hand in hand with the API key creation? And we delegate that to whomever is going to operate the app.

My proposal would be to create the usage plan(s) here, as part of the application, as they shouldn't change that often.

**Comment by kkiani** [comment]
@bramelfrink Usage plans are good addition. I totally agree with having usage plans. I think we can provide 3-4 predefined user plans to make life easier. If an end user needs more customize one they can create it then.

With regards to API Keys, The Component should provide at least one default API Key (with largest usage plan?) for the code owner in case they wants to interact with the model. The default could be shared on ssm parameter or bettter, aws secret manager so later the controller class can have access to it.

I did take a look at AWS Docs and noticed that it is not possible to create API Key after deployment. So, dynamic creation of API Keys after deployment is not possible anyway. However, the component should provide such interface for registering a list of generated API keys. Something like below:

```python
AwsVllmComponent(
    name="my-vllm",
    args=AwsVllmComponentArgs(
        region="eu-west-1",
        public_internet_access=True,
        api_keys=[
            APIKey(id_="my-api-key", value="api-key-value", usage_plan_tier=APIKey.developer_tier_usage_plan()),
        ]
    ),
)
```

File: src/damavand/cloud/aws/resources/vllm_component.py [hunk 1]
```diff
@@ -359,41 +335,150 @@ def api_resource_completions(self) -> aws.apigateway.Resource:
...
+        else:
+            return aws.apigateway.Method(
+                resource_name=f"{self._name}-api-method",
+                opts=ResourceOptions(parent=self),
+                rest_api=self.api.id,
+                resource_id=self.api_resource_completions.id,
+                http_method="POST",
+                authorization="NONE",
+                api_key_required=True,
+            )
```
Comment (kkiani) [inline src/damavand/cloud/aws/resources/vllm_component.py]: You can remove the if condition just by passing the `public_internet_access` to the `api_key_required` argument:

```suggestion
        return aws.apigateway.Method(
            resource_name=f"{self._name}-api-method",
            opts=ResourceOptions(parent=self),
            rest_api=self.api.id,
            resource_id=self.api_resource_completions.id,
            http_method="POST",
            authorization="NONE",
            api_key_required=self.args.public_internet_access,
        )
```

File: src/damavand/cloud/aws/resources/vllm_component.py [hunk 2]
```diff
@@ -125,19 +125,29 @@ def __init__(
...
-            _ = self.api_method_response
-            _ = self.api_deployment
-            _ = self.endpoint_ssm_parameter
+        _ = self.api
+        _ = self.api_resource_v1
+        _ = self.api_resource_chat
+        _ = self.api_resource_completions
+
+        # Only create API key if public internet access is set to False
+        if not self.args.public_internet_access:
```
Comment (kkiani) [inline src/damavand/cloud/aws/resources/vllm_component.py]: It looks like you are changing the behavior to always create the API Gateway which I do agree with but, change the arg name to something meaningful and related like this:

```suggestion
        if not self.args.is_api_key_required:
```

File: src/damavand/cloud/aws/resources/vllm_component.py [hunk 3]
```diff
@@ -359,41 +335,150 @@ def api_resource_completions(self) -> aws.apigateway.Resource:
...
+            resource_name=f"{self._name}-api-usage-plan",
+            opts=ResourceOptions(parent=self),
+            api_stages=[
+                aws.apigateway.UsagePlanApiStageArgs(
+                    api_id=self.api.id,
+                    # NOTE: How do we want to deal with API stages vs. AWS environments?
+                    stage=self.args.api_env_name,
+                )
+            ],
+        )
```
Comment (kkiani) [inline src/damavand/cloud/aws/resources/vllm_component.py]: Can we have a two or three other default usage plan?

File: src/damavand/cloud/aws/resources/vllm_component.py [hunk 4]
```diff
@@ -359,41 +354,226 @@ def api_resource_completions(self) -> aws.apigateway.Resource:
...
+            resource_name=f"{self._name}-tier-1-api-usage-plan",
+            opts=ResourceOptions(parent=self),
+            api_stages=[
+                aws.apigateway.UsagePlanApiStageArgs(
+                    api_id=self.api.id,
+                    stage=self.args.api_env_name,
+                )
+            ],
+            throttle_settings=aws.apigateway.UsagePlanThrottleSettingsArgs(
+                rate_limit=500
```
Comment (bramelfrink) [inline src/damavand/cloud/aws/resources/vllm_component.py:488]: @kkiani I'm only changing the rate limit at the moment. 

It's also possible to define quotas per day/week/month, but I think that's only interesting when you start pricing API's.

File: src/damavand/cloud/aws/resources/vllm_component.py [hunk 5]
```diff
@@ -29,8 +29,8 @@ class AwsVllmComponentArgs:
         number of instances to deploy the model.
     instance_type : str
         type of instance to deploy the model.
-    public_internet_access : bool
-        whether to deploy a public API for the model.
+    api_key_required : bool
+        whether an API key is will be required for interacting with the API.
```
Comment (kkiani) [inline src/damavand/cloud/aws/resources/vllm_component.py]:
```suggestion
        whether an API key is required for interacting with the API.
```

File: src/damavand/cloud/aws/resources/vllm_component.py (hunk 4 above)
Comment (kkiani) [reply src/damavand/cloud/aws/resources/vllm_component.py:488]: @bramelfrink, No I think `rate_limit` is enough.

**Review by kkiani — APPROVED** [review]

**Pull request merged**